apps/opik-backend/src/main/java/com/comet/opik/domain/ChatCompletionService.java

```diff
@@ -16,170 +14,139 @@
 import jakarta.ws.rs.ClientErrorException;
 import jakarta.ws.rs.InternalServerErrorException;
 import jakarta.ws.rs.ServerErrorException;
+import jakarta.ws.rs.core.Response;
 import lombok.NonNull;
 import lombok.extern.slf4j.Slf4j;
-import org.apache.commons.lang3.StringUtils;
+import org.apache.hc.core5.http.HttpStatus;
 import org.glassfish.jersey.server.ChunkedOutput;
 import ru.vyarus.dropwizard.guice.module.yaml.bind.Config;
 
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.util.Optional;
+import java.util.function.Consumer;
 
 @Singleton
 @Slf4j
 public class ChatCompletionService {
-
-    private static final String UNEXPECTED_ERROR_CALLING_LLM_PROVIDER = "Unexpected error calling LLM provider";
+    public static final String UNEXPECTED_ERROR_CALLING_LLM_PROVIDER = "Unexpected error calling LLM provider";
 
     private final LlmProviderClientConfig llmProviderClientConfig;
-    private final LlmProviderApiKeyService llmProviderApiKeyService;
+    private final LlmProviderFactory llmProviderFactory;
     private final RetryUtils.RetryPolicy retryPolicy;
 
     @Inject
     public ChatCompletionService(
-            @NonNull @Config LlmProviderClientConfig llmProviderClientConfig,
-            @NonNull LlmProviderApiKeyService llmProviderApiKeyService) {
-        this.llmProviderApiKeyService = llmProviderApiKeyService;
+            @NonNull @Config LlmProviderClientConfig llmProviderClientConfig, LlmProviderFactory llmProviderFactory) {
         this.llmProviderClientConfig = llmProviderClientConfig;
+        this.llmProviderFactory = llmProviderFactory;
         this.retryPolicy = newRetryPolicy();
     }
 
-    private RetryUtils.RetryPolicy newRetryPolicy() {
-        var retryPolicyBuilder = RetryUtils.retryPolicyBuilder();
-        Optional.ofNullable(llmProviderClientConfig.getMaxAttempts()).ifPresent(retryPolicyBuilder::maxAttempts);
-        Optional.ofNullable(llmProviderClientConfig.getJitterScale()).ifPresent(retryPolicyBuilder::jitterScale);
-        Optional.ofNullable(llmProviderClientConfig.getBackoffExp()).ifPresent(retryPolicyBuilder::backoffExp);
-        return retryPolicyBuilder
-                .delayMillis(llmProviderClientConfig.getDelayMillis())
-                .build();
-    }
-
     public ChatCompletionResponse create(@NonNull ChatCompletionRequest request, @NonNull String workspaceId) {
-        log.info("Creating chat completions, workspaceId '{}', model '{}'", workspaceId, request.model());
-        var openAiClient = getAndConfigureOpenAiClient(request, workspaceId);
+        var llmProviderClient = llmProviderFactory.getService(workspaceId, request.model());
+        llmProviderClient.validateRequest(request);
```

Comment:
This could be validated at the controller level

Reply:
This can be validated only after a provider specific client was obtained from the factory as this is meant for provider specific validations.
Therefore I believe it belongs in the service layer as it is right now.